src/nbl/video/COpenGL_Swapchain.cpp

```diff
@@ -0,0 +1,414 @@
+    COpenGLFence* glFen = IBackendObject::compatibility_cast<COpenGLFence*>(fence, this);
+    if (semaphore && !glSem)
+        return EAIR_ERROR;
+    if (fence && !glFen)
+        return EAIR_ERROR;
+
+    // TODO currently completely ignoring `timeout`
+
+    ++m_imgIx;
+    m_imgIx %= static_cast<uint32_t>(m_imageCount);
+
+    if (semaphore || fence)
+    {
+        core::smart_refctd_ptr<COpenGLSync> sync = m_threadHandler->getSyncForImgIx(m_imgIx);
+        if (glSem)
+            glSem->associateGLSync(core::smart_refctd_ptr(sync));
+        if (glFen)
+            glFen->associateGLSync(core::smart_refctd_ptr(sync));
+    }
+
+    assert(out_imgIx);
+    out_imgIx[0] = m_imgIx;
+
+    return EAIR_SUCCESS;
+}
+
+template <typename FunctionTableType_>
+nbl::video::ISwapchain::E_PRESENT_RESULT COpenGL_Swapchain<FunctionTableType_>::present(IGPUQueue* queue, const SPresentInfo& info)
+{
+    for (uint32_t i = 0u; i < info.waitSemaphoreCount; ++i)
+    {
+        if (getOriginDevice() != info.waitSemaphores[i]->getOriginDevice())
+            return ISwapchain::EPR_ERROR;
+    }
+
+    uint32_t _imgIx = info.imgIndex;
+    uint32_t semCount = info.waitSemaphoreCount;
+    IGPUSemaphore* const* const sems = info.waitSemaphores;
+
+    if (_imgIx >= m_params.minImageCount)
+        return ISwapchain::EPR_ERROR;
+    for (uint32_t i = 0u; i < semCount; ++i)
+    {
+        if (!this->isCompatibleDevicewise(sems[i]))
+            return ISwapchain::EPR_ERROR;
+    }
+    m_threadHandler->requestBlit(_imgIx, semCount, sems);
+
+    return ISwapchain::EPR_SUCCESS;
+}
+
+template <typename FunctionTableType_>
+core::smart_refctd_ptr<IGPUImage> COpenGL_Swapchain<FunctionTableType_>::createImage(const uint32_t imageIndex)
+{
+    if (!setImageExists(imageIndex))
+        return nullptr;
+
+    auto& device = m_threadHandler->m_device;
+    auto img_dst = core::smart_refctd_ptr_static_cast<COpenGLImage>(device->createImage(IGPUImage::SCreationParams(m_imgCreationParams)));
+    img_dst->m_optionalBackingSwapchain = core::smart_refctd_ptr<ISwapchain>(this);
+    img_dst->m_optionalIndexWithinSwapchain = imageIndex;
+
+    auto mreq = img_dst->getMemoryReqs();
+    mreq.memoryTypeBits &= device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+    auto imgMem = device->allocate(mreq, img_dst.get());
+    if (!img_dst || !imgMem.isValid())
+        return nullptr;
+
+    // weird validation that the original impl did
+    auto& ci = img_dst->getCreationParameters();
+    if (ci.type != asset::IImage::ET_2D)
+        return nullptr;
+    if (ci.arrayLayers != m_imgCreationParams.arrayLayers)
+        return nullptr;
+    if (ci.mipLevels != 1u)
+        return nullptr;
+    if (ci.extent.width != m_imgCreationParams.extent.width)
+        return nullptr;
+    if (ci.extent.height != m_imgCreationParams.extent.height)
+        return nullptr;
+    
+    m_threadHandler->createImgViewFbo(imageIndex, img_dst);
+
+    return img_dst;
+}
```

Comment:
rewrite to be closer to Vulkan

Reply:
retrieve image GLuint, don't create it, create an "imported" COpenGLImage

Reply:
no image view creation necessary, in GL image also as an implicit view associated with it